File: delphi/RayTracer.pas
unit RayTracer;

interface

uses Classes, Vcl.Graphics, Math, Generics.Collections, Windows, SysUtils;

type
  Surface = class;

  Vector = record
    x, y, z: Double;
    class operator Add(const a, b: Vector): Vector;
    class operator Subtract(const a, b: Vector): Vector;
    class operator Multiply(k: Double; const b: Vector): Vector;

    constructor Create(x, y, z: Double);

    function Dot(const v: Vector): Double;
    function Norm(): Vector;
    function Cross(const v: Vector): Vector;
    function Length(): Double;
  end;

  RGBColor = record
    b, g, r, a: Byte
  end;

  Color = record
    r, g, b: Double;

    class var White: Color;
    class var Grey: Color;
    class var Black: Color;
    class var Background: Color;
    class var DefaultColor: Color;

    constructor Create(r, g, b: Double);

    class operator Add(const a, b: Color): Color;
    class operator Multiply(k: Double; const b: Color): Color;
    class operator Multiply(const a, b: Color): Color;
    class function Clamp(v: Double): Byte; static;

    function ToDrawingColor: RGBColor;
  end;

  Camera = record
    ForwardDir: Vector;
    Right: Vector;
    Up: Vector;
    Pos: Vector;
    constructor Create(pos: Vector; lookAt: Vector);
  end;

  Thing = class;

  Ray = record
    Start: Vector;
    Dir: Vector;
    constructor Create(const start, dir: Vector);
  end;

  Intersection = record
    Thing: Thing;
    Ray: Ray;
    Dist: Double;
    constructor Create(Thing: Thing; const Ray: Ray; dist: Double);
    class function Invalid: Intersection; static;
    function IsValid: boolean;
  end;

  Thing = class(TObject)
  public
    function Intersect(const Ray: Ray): Intersection; virtual; abstract;
    function Normal(const pos: Vector): Vector; virtual; abstract;
    function Surface: Surface; virtual; abstract;
  end;

  Surface = class abstract
  public
    function Diffuse(const pos: Vector): Color; virtual; abstract;
    function Specular(const pos: Vector): Color; virtual; abstract;
    function Reflect(const pos: Vector): Double; virtual; abstract;
    function Roughness: Double; virtual; abstract;
  end;

  ShinySurface = class(Surface)
  public
    function Diffuse(const pos: Vector): Color; override;
    function Specular(const pos: Vector): Color; override;
    function Reflect(const pos: Vector): Double; override;
    function Roughness: Double; override;
  end;

  CheckerboardSurface = class(Surface)
  public
    function Diffuse(const pos: Vector): Color; override;
    function Specular(const pos: Vector): Color; override;
    function Reflect(const pos: Vector): Double; override;
    function Roughness: Double; override;
  end;

  Light = record
    Pos: Vector;
    Color: Color;
    constructor Create(pos: Vector; Color: Color);
  end;

  Scene = class(TObject)
  public
    things: array of Thing;
    lights: array of Light;
    xcamera: Camera;
    constructor Create;
  end;

  RayTracerEngine = class
  private
    MaxDepth: Integer;
    Scene: Scene;
  public
    constructor Create;
    function Intersections(const Ray: Ray): Intersection;
    function TraceRay(const Ray: Ray; depth: Integer): Color;
    function Shade(isect: Intersection; depth: Integer): Color;
    function GetReflectionColor(Thing: Thing; const pos, normal, rd: Vector;
      depth: Integer): Color;
    function GetNaturalColor(Thing: Thing; const pos, norm, rd: Vector): Color;
    procedure Render(Scene: Scene; img: Vcl.Graphics.TBitmap);
  end;

  Sphere = class(Thing)
  private
    radius2: Double;
    center: Vector;
    _surface: Surface;
  public
    function Intersect(const Ray: Ray): Intersection; override;
    function Normal(const pos: Vector): Vector; override;
    function Surface: Surface; override;
    constructor Create(center: Vector; radius: Double; Surface: Surface);
  end;

  Plane = class(Thing)
  private
    norm: Vector;
    offset: Double;
    _surface: Surface;
  public
    function Intersect(const Ray: Ray): Intersection; override;
    function Normal(const pos: Vector): Vector; override;
    function Surface: Surface; override;
    constructor Create(norm: Vector; offset: Double; Surface: Surface);
  end;

implementation

{ Vector }
constructor Vector.Create(x, y, z: Double);
begin
  self.x := x;
  self.y := y;
  self.z := z;
end;

function Vector.Cross(const v: Vector): Vector;
begin
  Result.x := y * v.z - z * v.y;
  Result.y := z * v.x - x * v.z;
  Result.z := x * v.y - y * v.x;
end;

function Vector.Dot(const v: Vector): Double;
begin
  Result := (x * v.x) + (y * v.y) + (z * v.z);
end;

function Vector.Length: Double;
begin
  Result := Sqrt(x * x + y * y + z * z);
end;

class operator Vector.Multiply(k: Double; const b: Vector): Vector;
begin
  Result.x := k * b.x;
  Result.y := k * b.y;
  Result.z := k * b.z;
end;

function Vector.Norm: Vector;
var
  divBy, m: Double;
begin
  m := self.Length();

  if (m = 0) then
    divBy := MaxDouble
  else
    divBy := 1.0 / m;

  Result := divBy * self;
end;

class operator Vector.Subtract(const a, b: Vector): Vector;
begin
  Result.x := a.x - b.x;
  Result.y := a.y - b.y;
  Result.z := a.z - b.z;
end;

class operator Vector.Add(const a, b: Vector): Vector;
begin
  Result.x := a.x + b.x;
  Result.y := a.y + b.y;
  Result.z := a.z + b.z;
end;

{ Color }

class operator Color.Add(const a, b: Color): Color;
begin
  Result.r := a.r + b.r;
  Result.g := a.g + b.g;
  Result.b := a.b + b.b;
end;

class operator Color.Multiply(k: Double; const b: Color): Color;
begin
  Result.r := k * b.r;
  Result.g := k * b.g;
  Result.b := k * b.b;
end;

class function Color.Clamp(v: Double): Byte;
begin
  if v > 255 then
    Result := 255
  else
    Result := Byte(Floor(v));
end;

constructor Color.Create(r, g, b: Double);
begin
  self.r := r;
  self.g := g;
  self.b := b;
end;

class operator Color.Multiply(const a, b: Color): Color;
begin
  Result.r := a.r * b.r;
  Result.g := a.g * b.g;
  Result.b := a.b * b.b;
end;

function Color.ToDrawingColor: RGBColor;
begin
  Result.r := Color.Clamp(self.r * 255.0);
  Result.g := Color.Clamp(self.g * 255.0);
  Result.b := Color.Clamp(self.b * 255.0);
  Result.a := 255;
end;

{ Camera }

constructor Camera.Create(pos, lookAt: Vector);
var
  down: Vector;
begin

  down := Vector.Create(0.0, -1.0, 0.0);

  self.Pos := pos;
  self.ForwardDir := (lookAt - self.Pos).Norm;
  self.Right := 1.5 * self.ForwardDir.Cross(down).Norm();
  self.Up := 1.5 * self.ForwardDir.Cross(self.Right).Norm();
end;

{ Ray }

constructor Ray.Create(const start, dir: Vector);
begin
  self.Start := start;
  self.Dir := dir;
end;

{ Intersection }

constructor Intersection.Create(Thing: Thing; const Ray: Ray; dist: Double);
begin
  self.Thing := Thing;
  self.Ray := Ray;
  self.Dist := dist;
end;

class function Intersection.Invalid: Intersection;
begin
  Result.Thing := nil;
  Result.Dist := 0;
end;

function Intersection.IsValid: boolean;
begin
  Result := Assigned(self.Thing);
end;

{ Light }
constructor Light.Create(pos: Vector; Color: Color);
begin
  self.Pos := pos;
  self.Color := Color;
end;

{ Sphere }
constructor Sphere.Create(center: Vector; radius: Double; Surface: Surface);
begin
  self.radius2 := radius * radius;
  self.center := center;
  self._surface := Surface;
end;

function Sphere.Intersect(const Ray: Ray): Intersection;
var
  eo: Vector;
  v, dist, disc: Double;
begin

  eo := self.center - Ray.Start;
  v := eo.Dot(Ray.Dir);
  dist := 0;

  if (v >= 0) then
  begin
    disc := self.radius2 - (eo.Dot(eo) - (v * v));
    if (disc >= 0) then
    begin
      dist := v - Sqrt(disc);
    end;
  end;

  if (dist = 0) then
    Result := Intersection.Invalid()
  else
  begin
    Result := Intersection.Create(self, Ray, dist);
  end;

end;

function Sphere.Normal(const pos: Vector): Vector;
begin
  Result := (pos - self.center).Norm;
end;

function Sphere.Surface: Surface;
begin
  Result := self._surface;
end;

{ ShinySurface }
function ShinySurface.Diffuse(const pos: Vector): Color;
begin
  Result := Color.White;
end;

function ShinySurface.Reflect(const pos: Vector): Double;
begin
  Result := 0.7;
end;

function ShinySurface.Roughness: Double;
begin
  Result := 250;
end;

function ShinySurface.Specular(const pos: Vector): Color;
begin
  Result := Color.Grey;
end;

{ CheckerboardSurface }
function CheckerboardSurface.Diffuse(const pos: Vector): Color;
begin
  if ((Floor(pos.z) + Floor(pos.x)) mod 2) <> 0 then
    Result := Color.White
  else
    Result := Color.Black;
end;

function CheckerboardSurface.Reflect(const pos: Vector): Double;
begin
  if ((Floor(pos.z) + Floor(pos.x)) mod 2) <> 0 then
    Result := 0.1
  else
    Result := 0.7;
end;

function CheckerboardSurface.Roughness: Double;
begin
  Result := 150.0;
end;

function CheckerboardSurface.Specular(const pos: Vector): Color;
begin
  Result := Color.White;
end;

{ Plane }

constructor Plane.Create(norm: Vector; offset: Double; Surface: Surface);
begin
  self.norm := norm;
  self.offset := offset;
  self._surface := Surface;
end;

function Plane.Intersect(const Ray: Ray): Intersection;
var
  dist, denom: Double;
begin
  denom := self.norm.Dot(Ray.Dir);
  if (denom > 0) then
  begin
    Result := Intersection.Invalid();
  end else
  begin
    dist := (norm.Dot(Ray.Start) + offset) / (-denom);
    Result := Intersection.Create(self, Ray, dist);
  end;
end;

function Plane.Normal(const pos: Vector): Vector;
begin
  Result := self.norm;
end;

function Plane.Surface: Surface;
begin
  Result := _surface;
end;

{ RayTracerEngine }
constructor RayTracerEngine.Create;
begin
  self.MaxDepth := 5;
end;

function RayTracerEngine.GetNaturalColor(Thing: Thing;
  const pos, norm, rd: Vector): Color;
var
  diffuseColor, specularColor: Color;
  ldis, livec: Vector;
  testRay: Ray;
  neatIsect: Intersection;
  inShadow: boolean;
  illum, specular: Double;
  lcolor, scolor: Color;
  item: Light;
  roughness: Double;
begin

  Result := Color.DefaultColor;
  diffuseColor := Thing.Surface.Diffuse(pos);
  specularColor := Thing.Surface.Specular(pos);
  roughness := Thing.Surface.Roughness;
  testRay.Start := pos;

  for item in Scene.lights do
  begin
    ldis := item.Pos - pos;
    livec := ldis.Norm;

    testRay.Dir := livec;
    neatIsect := self.Intersections(testRay);

    inShadow := neatIsect.IsValid() and (neatIsect.Dist <= ldis.Length());

    if not inShadow then
    begin
      illum := livec.Dot(norm);
      specular := livec.Dot(rd.Norm);

      if illum > 0 then
        lcolor := illum * item.Color
      else
        lcolor := Color.DefaultColor;

      if specular > 0 then
        scolor := Math.Power(specular, roughness) * item.Color
      else
        scolor := Color.DefaultColor;

      Result := Result + ((diffuseColor * lcolor) + (specularColor * scolor));
    end;
  end;

end;

function RayTracerEngine.GetReflectionColor(Thing: Thing;
  const pos, normal, rd: Vector; depth: Integer): Color;
var
  r: Ray;
begin
  r := Ray.Create(pos, rd);
  Result := Thing.Surface.Reflect(pos) * self.TraceRay(r, depth + 1);
end;

function RayTracerEngine.Intersections(const Ray: Ray): Intersection;
var
  closest: Double;
  inter, closestInter: Intersection;
  i: Integer;
begin
  closest := MaxInt;
  closestInter := Intersection.Invalid();

  for i := 0 to High(Scene.things) do
  begin
    inter := Scene.things[i].Intersect(Ray);
    if (inter.IsValid()) and (inter.Dist < closest) then
    begin
      closestInter := inter;
      closest := inter.Dist;
    end;
  end;
  Result := closestInter;
end;

procedure RayTracerEngine.Render(Scene: Scene; img: Vcl.Graphics.TBitmap);
var
  x, y, w, h: Integer;
  destColor: Color;
  testRay: Ray;
  c: RGBColor;
  stride, pos: Integer;
  start: PBYTE;

  function getPoint(x, y: Integer; Camera: Camera): Vector;
  var
    recenterX, recenterY: Double;
  begin
    recenterX := (x - (w / 2.0)) / 2.0 / w;
    recenterY := -(y - (h / 2.0)) / 2.0 / h;
    Result := (Camera.ForwardDir + ((recenterX * Camera.Right) +
      (recenterY * Camera.Up))).Norm();
  end;

begin

  self.Scene := Scene;
  w := img.Width - 1;
  h := img.Height - 1;

  start := img.ScanLine[0];
  stride := Integer(img.ScanLine[1]) - Integer(img.ScanLine[0]);

  for y := 0 to h do
  begin
    pos := y * stride;
    for x := 0 to w do
    begin
      testRay.Start := Scene.xcamera.Pos;
      testRay.Dir := getPoint(x, y, Scene.xcamera);

      destColor := self.TraceRay(testRay, 0);

      c := destColor.ToDrawingColor;
      start[pos] := c.b;
      start[pos + 1] := c.g;
      start[pos + 2] := c.r;
      start[pos + 3] := 255;
      pos := pos + 4;
    end;
  end;
end;

function RayTracerEngine.Shade(isect: Intersection; depth: Integer): Color;
var
  d: Vector;
  pos, normal, reflectDir: Vector;
  naturalColor, reflectedColor: Color;

begin
  d := isect.Ray.Dir;
  pos := (isect.Dist * d) + isect.Ray.Start;
  normal := isect.Thing.Normal(pos);
  reflectDir := d - (2.0 * normal.Dot(d) * normal);

  naturalColor := Color.Background + self.GetNaturalColor(isect.Thing, pos,
    normal, reflectDir);

  if depth >= self.MaxDepth then
    reflectedColor := Color.Grey
  else
    reflectedColor := self.GetReflectionColor(isect.Thing, pos, normal,
      reflectDir, depth);

  Result := naturalColor + reflectedColor;
end;

function RayTracerEngine.TraceRay(const Ray: Ray; depth: Integer): Color;
var
  isect: Intersection;
begin
  isect := self.Intersections(Ray);
  if (isect.IsValid()) then
  begin
    Result := self.Shade(isect, depth);
  end else
  begin
    Result := Color.Background;
  end;
end;

{ Scene }
constructor Scene.Create;
var
  shiny: ShinySurface;
  checkerboard: CheckerboardSurface;
begin

  SetLength(things, 3);
  SetLength(lights, 4);

  shiny := ShinySurface.Create;
  checkerboard := CheckerboardSurface.Create;

  things[0] := (Plane.Create(Vector.Create(0.0, 1.0, 0.0), 0.0,    checkerboard));
  things[1] := (Sphere.Create(Vector.Create(0.0, 1.0, -0.25), 1.0, shiny));
  things[2] := (Sphere.Create(Vector.Create(-1.0, 0.5, 1.5), 0.5,  shiny));

  lights[0] := (Light.Create(Vector.Create(-2.0, 2.5, 0.0), Color.Create(0.49, 0.07, 0.07)));
  lights[1] := (Light.Create(Vector.Create(1.5, 2.5, 1.5),  Color.Create(0.07,  0.07, 0.49)));
  lights[2] := (Light.Create(Vector.Create(1.5, 2.5, -1.5), Color.Create(0.07, 0.49, 0.071)));
  lights[3] := (Light.Create(Vector.Create(0.0, 3.5, 0.0),  Color.Create(0.21,  0.21, 0.35)));

  self.xcamera := Camera.Create(Vector.Create(3.0, 2.0, 4.0), Vector.Create(-1.0, 0.5, 0.0));
end;

initialization

Color.White := Color.Create(1, 1, 1);
Color.Grey := Color.Create(0.5, 0.5, 0.5);
Color.Black := Color.Create(0, 0, 0);

Color.DefaultColor := Color.Black;
Color.Background := Color.Black;

end.

File: rust/src/ray_tracer.rs
#![allow(unused_must_use, dead_code)]

extern crate bmp;
extern crate rayon;
use bmp::Image;
use bmp::Pixel;
use std::time::Instant;

use std::{env, ops};

use rayon::prelude::*;

const FAR_AWAY: f32 = 1000000.0;

#[derive(Debug, Copy, Clone)]
struct RgbColor {
    b: u8,
    g: u8,
    r: u8,
    a: u8,
}

#[derive(Debug, Copy, Clone)]
struct Vector {
    x: f32,
    y: f32,
    z: f32,
}

impl Vector {
    fn new(x: f32, y: f32, z: f32) -> Vector {
        Vector { x, y, z }
    }

    fn mag(&self) -> f32 {
        let value = self.x * self.x + self.y * self.y + self.z * self.z;
        value.sqrt()
    }

    fn norm(&self) -> Vector {
        let mag = self.mag();
        let div = if mag == 0.0 { FAR_AWAY } else { 1.0 / mag };
        *self * div
    }

    fn dot(&self, v: Vector) -> f32 {
        self.x * v.x + self.y * v.y + self.z * v.z
    }
}

impl ops::Add<Vector> for Vector {
    type Output = Self;

    fn add(self, v: Vector) -> Self {
        Self::new(self.x + v.x, self.y + v.y, self.z + v.z)
    }
}

impl ops::Sub<Vector> for Vector {
    type Output = Self;

    fn sub(self, v: Vector) -> Self {
        Self::new(self.x - v.x, self.y - v.y, self.z - v.z)
    }
}

impl ops::Mul<Vector> for Vector {
    type Output = Self;

    fn mul(self, v: Self) -> Self {
        Self::new(
            self.y * v.z - self.z * v.y,
            self.z * v.x - self.x * v.z,
            self.x * v.y - self.y * v.x,
        )
    }
}

impl ops::Mul<f32> for Vector {
    type Output = Self;

    fn mul(self, k: f32) -> Self {
        Self::new(k * self.x, k * self.y, k * self.z)
    }
}

#[derive(Debug, Copy, Clone)]
struct Color {
    r: f32,
    g: f32,
    b: f32,
}

const COLOR_WHITE: Color = Color {
    r: 1.0,
    g: 1.0,
    b: 1.0,
};
const COLOR_GREY: Color = Color {
    r: 0.5,
    g: 0.5,
    b: 0.5,
};
const COLOR_BLACK: Color = Color {
    r: 0.0,
    g: 0.0,
    b: 0.0,
};
const COLOR_BACKGROUND: Color = COLOR_BLACK;
const COLOR_DEFAULT_COLOR: Color = COLOR_BLACK;

impl Color {
    fn new(r: f32, g: f32, b: f32) -> Color {
        Color { r, g, b }
    }

    fn to_drawing_color(&self) -> RgbColor {
        RgbColor {
            r: (self.r.clamp(0.0, 1.0) * 255.0) as u8,
            g: (self.g.clamp(0.0, 1.0) * 255.0) as u8,
            b: (self.b.clamp(0.0, 1.0) * 255.0) as u8,
            a: 255,
        }
    }
}

impl ops::Mul<Color> for Color {
    type Output = Self;
    fn mul(self, c: Color) -> Self {
        Self::new(self.r * c.r, self.g * c.g, self.b * c.b)
    }
}

impl ops::Mul<f32> for Color {
    type Output = Self;
    fn mul(self, k: f32) -> Self {
        Self::new(k * self.r, k * self.g, k * self.b)
    }
}

impl ops::Add<Color> for Color {
    type Output = Self;
    fn add(self, c: Color) -> Self {
        Self::new(self.r + c.r, self.g + c.g, self.b + c.b)
    }
}

impl ops::AddAssign<Color> for Color {
    fn add_assign(&mut self, other: Self) {
        *self = *self + other;
    }
}

#[derive(Debug, Copy, Clone)]
struct Camera {
    forward: Vector,
    right: Vector,
    up: Vector,
    pos: Vector,
}

impl Camera {
    fn new(pos: Vector, look_at: Vector) -> Camera {
        let down = Vector::new(0.0, -1.0, 0.0);
        let forward = (look_at - pos).norm();
        let right = (forward * down).norm() * 1.5;

        Camera {
            pos,
            forward,
            right,
            up: (forward * right).norm() * 1.5,
        }
    }
}

#[derive(Debug, Copy, Clone)]
struct Ray {
    start: Vector,
    dir: Vector,
}

#[derive(Debug, Copy, Clone)]
struct Intersection {
    thing: Thing,
    ray: Ray,
    dist: f32,
}

impl Intersection {
    fn new(thing: Thing, ray: Ray, dist: f32) -> Intersection {
        Intersection { thing, ray, dist }
    }
}

#[derive(Debug, Copy, Clone)]
enum Surface {
    CheckerboardSurface,
    ShinySurface,
}

struct SurfaceProperties {
    diffuse: Color,
    specular: Color,
    reflect: f32,
    roughness: f32,
}

impl Surface {
    fn get_properties(&self, pos: Vector) -> SurfaceProperties {
        match self {
            Surface::CheckerboardSurface => {
                let (diffuse, reflect) = if (pos.z.floor() + pos.x.floor()) as i32 % 2 != 0 {
                    (COLOR_WHITE, 0.1)
                } else {
                    (COLOR_BLACK, 0.7)
                };

                SurfaceProperties {
                    diffuse,
                    specular: COLOR_WHITE,
                    reflect,
                    roughness: 150.0,
                }
            }
            Surface::ShinySurface => SurfaceProperties {
                diffuse: COLOR_WHITE,
                specular: COLOR_GREY,
                reflect: 0.7,
                roughness: 250.0,
            },
        }
    }
}

#[derive(Debug, Copy, Clone)]
struct SphereInfo {
    surface: Surface,
    radius2: f32,
    center: Vector,
}

#[derive(Debug, Copy, Clone)]
struct PlaneInfo {
    surface: Surface,
    offset: f32,
    normal: Vector,
}

#[derive(Debug, Copy, Clone)]
enum Thing {
    Plane(PlaneInfo),
    Sphere(SphereInfo),
}

struct Light {
    pos: Vector,
    color: Color,
}

impl Thing {
    fn new_sphere(center: Vector, radius: f32, surface: Surface) -> Thing {
        Thing::Sphere(SphereInfo {
            surface,
            radius2: radius * radius,
            center,
        })
    }

    fn new_plane(normal: Vector, offset: f32, surface: Surface) -> Thing {
        Thing::Plane(PlaneInfo {
            surface,
            offset,
            normal,
        })
    }

    fn normal(&self, pos: Vector) -> Vector {
        match self {
            Thing::Sphere(ref sphere) => (pos - sphere.center).norm(),
            Thing::Plane(ref plane) => plane.normal,
        }
    }

    fn surface(&self) -> Surface {
        match self {
            Thing::Sphere(ref sphere) => sphere.surface,
            Thing::Plane(ref plane) => plane.surface,
        }
    }

    fn intersect(&self, ray: &Ray) -> Option<Intersection> {
        match self {
            Thing::Sphere(ref sphere) => {
                let eo = sphere.center - ray.start;
                let v = eo.dot(ray.dir);

                if v >= 0.0 {
                    let disc = sphere.radius2 - ((eo.dot(eo)) - (v * v));
                    if disc >= 0.0 {
                        let dist = v - disc.sqrt();
                        return Some(Intersection::new(*self, *ray, dist));
                    }
                }
                None
            }

            Thing::Plane(ref plane) => {
                let denom = plane.normal.dot(ray.dir);
                if denom > 0.0 {
                    return None;
                }
                let dist = (plane.normal.dot(ray.start) + plane.offset) / (-denom);
                Some(Intersection::new(*self, *ray, dist))
            }
        }
    }
}

struct Scene {
    things: Vec<Thing>,
    lights: Vec<Light>,
    camera: Camera,
}

impl Scene {
    fn new() -> Scene {
        Scene {
            things: vec![
                Thing::new_plane(
                    Vector::new(0.0, 1.0, 0.0),
                    0.0,
                    Surface::CheckerboardSurface,
                ),
                Thing::new_sphere(Vector::new(0.0, 1.0, -0.25), 1.0, Surface::ShinySurface),
                Thing::new_sphere(Vector::new(-1.0, 0.5, 1.5), 0.5, Surface::ShinySurface),
            ],
            lights: vec![
                Light {
                    pos: Vector::new(-2.0, 2.5, 0.0),
                    color: Color::new(0.49, 0.07, 0.07),
                },
                Light {
                    pos: Vector::new(1.5, 2.5, 1.5),
                    color: Color::new(0.07, 0.07, 0.49),
                },
                Light {
                    pos: Vector::new(1.5, 2.5, -1.5),
                    color: Color::new(0.07, 0.49, 0.071),
                },
                Light {
                    pos: Vector::new(0.0, 3.5, 0.0),
                    color: Color::new(0.21, 0.21, 0.35),
                },
            ],
            camera: Camera::new(Vector::new(3.0, 2.0, 4.0), Vector::new(-1.0, 0.5, 0.0)),
        }
    }
}

struct RayTracerEngine {
    max_depth: i32,
    scene: Scene,
}

impl RayTracerEngine {
    fn intersections(&self, ray: &Ray) -> Option<Intersection> {
        let mut closest = FAR_AWAY;
        let mut closest_intersection: Option<Intersection> = None;

        for thing in &self.scene.things {
            let inter = thing.intersect(ray);

            if let Some(result) = inter {
                if result.dist < closest {
                    closest_intersection = inter;
                    closest = result.dist;
                }
            }
        }
        closest_intersection
    }

    fn test_ray(&self, ray: &Ray) -> Option<f32> {
        self.intersections(ray).map(|r| r.dist)
    }

    fn trace_ray(&self, ray: &Ray, depth: i32) -> Color {
        self.intersections(ray)
            .map_or(COLOR_BACKGROUND, |r| self.shade(r, depth))
    }

    fn shade(&self, isect: Intersection, depth: i32) -> Color {
        let d: Vector = isect.ray.dir;
        let pos: Vector = (d * isect.dist) + isect.ray.start;
        let normal: Vector = isect.thing.normal(pos);
        let reflect_dir: Vector = d - (normal * normal.dot(d) * 2.0);

        let surface = isect.thing.surface().get_properties(pos);

        let natural_color =
            COLOR_BACKGROUND + self.get_natural_color(&surface, pos, normal, reflect_dir);
        let reflected_color = if depth >= self.max_depth {
            COLOR_GREY
        } else {
            self.get_reflection_color(&surface, pos, reflect_dir, depth)
        };

        natural_color + reflected_color
    }

    fn get_reflection_color(
        &self,
        surface: &SurfaceProperties,
        pos: Vector,
        rd: Vector,
        depth: i32,
    ) -> Color {
        let ray = Ray {
            start: pos,
            dir: rd,
        };
        self.trace_ray(&ray, depth + 1) * surface.reflect
    }

    fn get_natural_color(
        &self,
        surface: &SurfaceProperties,
        pos: Vector,
        norm: Vector,
        rd: Vector,
    ) -> Color {
        let mut result = COLOR_BLACK;
        let rd_norm = rd.norm();

        for light in &self.scene.lights {
            let ldis = light.pos - pos;
            let livec = ldis.norm();
            let ray = Ray {
                start: pos,
                dir: livec,
            };

            let is_in_shadow = if let Some(value) = self.test_ray(&ray) {
                value <= ldis.mag()
            } else {
                false
            };

            if !is_in_shadow {
                let illum = livec.dot(norm);
                let specular = livec.dot(rd_norm);

                let lcolor = if illum > 0.0 {
                    light.color * illum
                } else {
                    COLOR_DEFAULT_COLOR
                };

                let scolor = if specular > 0.0 {
                    light.color * specular.powf(surface.roughness)
                } else {
                    COLOR_DEFAULT_COLOR
                };

                result += (lcolor * surface.diffuse) + (scolor * surface.specular);
            }
        }
        result
    }

    pub fn get_point(
        &self,
        x: u32,
        y: u32,
        camera: Camera,
        screen_width: u32,
        screen_height: u32,
    ) -> Vector {
        let rx = (x as f32 - (screen_width as f32 / 2.0)) / 2.0 / screen_width as f32;
        let ry = -(y as f32 - (screen_height as f32 / 2.0)) / 2.0 / screen_height as f32;
        (camera.forward + (camera.right * rx) + (camera.up * ry)).norm()
    }

    pub fn render(&self, image: &mut Image, w: u32, h: u32) {
        let mut ray = Ray {
            start: self.scene.camera.pos,
            dir: Vector::new(0.0, 0.0, 0.0),
        };

        for y in 0..h {
            for x in 0..w {
                ray.dir = self.get_point(x, y, self.scene.camera, h, w);
                let color = self.trace_ray(&ray, 0).to_drawing_color();
                image.set_pixel(x, y, Pixel::new(color.r, color.g, color.b));
            }
        }
    }

    pub fn parallel_render(&self, image: &mut Image, w: u32, h: u32) {
        let data: Vec<(u32, u32, Pixel)> = (0..(h * w))
            .into_par_iter()
            .map(|v| (v / w, v % w))
            .map(|(y, x)| {
                let ray = Ray {
                    start: self.scene.camera.pos,
                    dir: self.get_point(x, y, self.scene.camera, h, w),
                };
                let color = self.trace_ray(&ray, 0).to_drawing_color();

                (y, x, Pixel::new(color.r, color.g, color.b))
            })
            .collect();

        for (y, x, p) in data {
            image.set_pixel(x, y, p);
        }
    }
}

fn main() {
    let is_parallel = env::args_os()
        .nth(1)
        .and_then(|s| s.into_string().map(|s| s == "parallel").ok())
        .unwrap_or(false);

    let width: u32 = 500;
    let height: u32 = 500;

    let mut image = Image::new(width, height);

    let n = 100;
    let engine = RayTracerEngine {
        max_depth: 5,
        scene: Scene::new(),
    };

    let now = Instant::now();

    if is_parallel {
        println!("Parallel!");

        for _ in 0..n {
            engine.parallel_render(&mut image, width, height);
        }
    } else {
        println!("Not parallel!");

        for _ in 0..n {
            engine.render(&mut image, width, height);
        }
    }

    let t = now.elapsed().as_millis();

    println!(
        "total time for {:?} iterations = {:?} ms, avg time = {:?} ms",
        n,
        t,
        t / n
    );

    image.save("RayTracer.bmp");
}
